libraries/common/stream.effekt

```diff
@@ -282,6 +293,18 @@ def source[A, R] { stream: () => Unit / emit[A] } { reader: () => R / read[A] }:
   }
 }
 
+/// Combines two streams together producing a stream of pairs in lockstep.
+/// Given two streams of length `n` and `m`, it produces a stream of length `min(n, m)`.
+def zip[A, B] { stream1: () => Unit / emit[A] } { stream2: () => Unit / emit[B] }: Unit / emit[(A, B)] = {
+  with source[A, Unit] { stream1 }
+  with source[B, Unit] { stream2 }
+
+  exhaustively {
+    val a = do read[A]()
+    val b = do read[B]()
+    do emit((a, b))
+  }
+}
```

Comment:
What's the idiomatic way to write `zip` and use it?

1. There's _something_ like `exhaustively` being used, but without the `action` parameter (added temporarily as a separate overload). What would be the idiomatic way to do this kind of thing?
2. I also don't really like that I have to specify the type params both in `for` and in `zip` when I use it. Would it instead make sense to make a `for` wrapper taking two streams? My intent is to mostly use it for iteration anyways...
3. Should we return a `These` instead? (`Left a | Right b | Both (a, b)`)